File: src/MultiAgent.Core/Models/Device.cs
namespace MultiAgent.Core.Models;

/// <summary>
/// Represents a single physical device (e.g. an ESP32 module) managed by the orchestrator.
/// </summary>
public sealed class Device
{
    public string DeviceId { get; init; }
    public string Name { get; init; }
    public string? IpAddress { get; set; }
    public string? MacAddress { get; init; }
    public DeviceCapability Capabilities { get; init; }
    public DeviceStatus Status { get; set; } = DeviceStatus.Unknown;
    public string FirmwareVersion { get; set; } = "0.0.0";
    public double CurrentFrequencyHz { get; set; } = 2_400_000_000.0;
    public int? Rssi { get; set; }
    public DateTimeOffset? LastSeen { get; set; }
    public Dictionary<string, object> Telemetry { get; } = [];

    public Device(string deviceId, string name,
                  string? ipAddress = null, string? macAddress = null,
                  DeviceCapability capabilities = DeviceCapability.Wifi | DeviceCapability.Ble)
    {
        DeviceId = deviceId;
        Name = name;
        IpAddress = ipAddress;
        MacAddress = macAddress;
        Capabilities = capabilities;
    }

    public bool HasCapability(DeviceCapability cap) => (Capabilities & cap) == cap;

    public void UpdateTelemetry(IReadOnlyDictionary<string, object> data)
    {
        foreach (var kv in data)
            Telemetry[kv.Key] = kv.Value;
        LastSeen = DateTimeOffset.UtcNow;
        if (data.TryGetValue("rssi", out var rssi) && rssi is int r) Rssi = r;
        if (data.TryGetValue("frequency_hz", out var freq) && freq is double f) CurrentFrequencyHz = f;
    }
}

File: ios-integration/Sources/IOSIntegration/Models/Device.swift
import Foundation

/// Represents a single ESP32 node managed by the Multi-Agent orchestrator.
public struct Device: Codable, Identifiable, Sendable {
    public let id: String
    public var name: String
    public var status: DeviceStatus
    public var firmwareVersion: String?
    public var lastSeen: Date?
    public var ipAddress: String?
    public var macAddress: String?
    public var capabilities: [String]

    public enum DeviceStatus: String, Codable, Sendable {
        case online
        case offline
        case busy
        case error
    }

    private enum CodingKeys: String, CodingKey {
        case id
        case name
        case status
        case firmwareVersion  = "firmware_version"
        case lastSeen         = "last_seen"
        case ipAddress        = "ip_address"
        case macAddress       = "mac_address"
        case capabilities
    }

    public init(
        id: String,
        name: String,
        status: DeviceStatus = .offline,
        firmwareVersion: String? = nil,
        lastSeen: Date? = nil,
        ipAddress: String? = nil,
        macAddress: String? = nil,
        capabilities: [String] = []
    ) {
        self.id = id
        self.name = name
        self.status = status
        self.firmwareVersion = firmwareVersion
        self.lastSeen = lastSeen
        self.ipAddress = ipAddress
        self.macAddress = macAddress
        self.capabilities = capabilities
    }
}

/// Payload used when registering a new device with the orchestrator.
public struct DeviceCreate: Codable, Sendable {
    public let name: String
    public let macAddress: String?
    public let capabilities: [String]

    private enum CodingKeys: String, CodingKey {
        case name
        case macAddress  = "mac_address"
        case capabilities
    }

    public init(name: String, macAddress: String? = nil, capabilities: [String] = []) {
        self.name = name
        self.macAddress = macAddress
        self.capabilities = capabilities
    }
}
